### PR TITLE
[gardenlet] Refactor handling of `Shoot` main controller special cases

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity quality
/kind technical-debt

**What this PR does / why we need it**:

Previously, special cases for confined and jittered shoot reconciliations (see [docs](https://github.com/gardener/gardener/blob/master/docs/concepts/gardenlet.md#shoot-controller)) were handled by a few lines of boolean variable combinations, and the reconciliation due tracker.
We faced some bugs and regressions in this logic because it was not covered by tests at all, was not readable, and was not documented at all.
However, this logic is quite important for the stability of large-scale gardener installations.

This PR refactors this logic to make it more readable and well-documented and adds full unit test coverage.
Additionally, we can now use the helpers in both the shoot main reconciler and its event handler.
This allows us to use `AddAfter` in the event handler, instead of performing a reconciliation that will exit early with `RequeueAfter` (and allows us to get rid of the reconciliation due tracker).

Co-Authored-By @rfranzke 

**Which issue(s) this PR fixes**:
Follow-up on https://github.com/gardener/gardener/pull/7082
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by timebertt** [comment]
This PR should go into the v1.62.0 release. Otherwise, the `reconcileInMaintenanceOnly=false` case is partially broken.
cc @oliver-goetz 
/milestone v1.62

File: pkg/gardenlet/controller/shoot/shoot/helper/infos.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	// if the shoot is failed or ignored, we don't want to requeue the shoot
+	if i.isFailed || i.isIgnored {
+		return reconcile.Result{}
+	}
+
+	// from here on, we have a case of regular reconciliation (i.e., shoot is observed at latest generation and in
+	// 'Reconcile Succeeded' state).
+
+	// If we do not confine reconciliations (either by the operator or shoot owner) to the maintenance time window then
+	// we allow immediate reconciliations.
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/helper/infos.go]:
```suggestion
	// we requeue according to the sync period configured in gardenlet configuration (which might have been overwritten on the shoot via annotation).
```

**Review by rfranzke — COMMENTED** [review]
Thanks, we did a joint review with @ary1992

**Review by rfranzke — COMMENTED** [review]
Thank you @timebertt for the joint work on this, enjoyed it.

/lgtm
/approve

**Comment by timebertt** [comment]
Thanks for the fast review, PTAL :)

**Comment by timuthy** [comment]
/unassign
for now since there are enough reviewers and I won't make it in time.

File: pkg/controllermanager/controller/shoot/maintenance/add.go [hunk 2]
```diff
@@ -15,6 +15,8 @@
 package maintenance
 
 import (
+	"k8s.io/utils/clock"
```
Comment (acumino) [inline pkg/controllermanager/controller/shoot/maintenance/add.go]:
```suggestion
```

**Review by acumino — CHANGES_REQUESTED** [review]

**Comment by timebertt** [comment]
Sorry folks, something went wrong during the rebase. It should be fixed now.

**Review by acumino — COMMENTED** [review]
/lgtm

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Comment by timuthy** [comment]
Kudos @timebertt and @rfranzke for touching this spaceship 👨‍🚀 Nice PR!

**Pull request merged**